sql_insert_writer/sql_insert_writer.py

```diff
@@ -0,0 +1,241 @@
+             ORDER BY ordinal_position'''
+
+    return db.query(qry, table_name=table_name).all()
+
+
+def col_data_sqlite(db, table_name):
+    """Gets metadata for a SQLite table's columns"""
+
+    qry = "PRAGMA TABLE_INFO('{}')".format(table_name)
+    # Alas cannot use proper parameters here; not recognized in
+    # the context of a PRAGMA statement
+
+    db.query(qry)
+    curs = db.db.connection.cursor()
+    try:
+        curs.execute(qry)
+        return [AttrDict({'table_name': table_name,
+                          'column_name': row.name,
+                          'data_type': row.type}) for row in db.query(qry)]
+    except ResourceClosedError:  # thus SQLite reacts to nonexistent table
+        return []
+
+
+col_data_functions = {
+    'postgresql': col_data_info_schema,
+    'sqlite': col_data_sqlite,
+    'mysql': col_data_info_schema,
+}
+
+
+def col_data(db, table_name):
+    """Gets metadata for a table's columns
+
+    Returns list, each element having .table, name, .column_name, .data_type attributes"""
+
+    db_type = db_engine_name(db.db_url)
+    col_data_function = col_data_functions.get(db_type)
+    if col_data_function:
+        result = col_data_function(db, table_name)
+    else:
+        raise NotImplementedError('{} not supported'.format(db_type))
+
+    if len(result) == 0:
+        raise BadDBNameError('No table {} in database'.format(table_name))
+
+    return result
+
+
+INSERT_TEMPLATE = '''
+INSERT INTO {destination} (
+{dest_column_block}
+)
+SELECT
+{source_column_block}
+FROM {from_clause}'''
+
+
+def remove_last(target, remove_me):
+    """
+    Returns string `target` with last occurance of `remove_me` removed
+
+    >>> remove_last('evacuate valorous vampires', 'va')
+    'evacuate valorous mpires'
+    """
+
+    before = target[:target.rfind(remove_me)]
+    after = target[target.rfind(remove_me) + len(remove_me):]
+    return before + after
+
+
+def no_value(db_url):
+    """
+    Value to insert when no value known; depends on database engine
+    """
+    if db_engine_name(db_url) == 'postgresql':
+        return 'DEFAULT'
+    else:
+        return 'NULL'
+
+
+INDENT = ' ' * 2
+
+
+def build_from_clause(sources):
+    """Given a list of table names, connects them with JOINs"""
+
+    from_clause = [sources[0]]
+    for join_to in sources[1:]:
+        from_clause.append('JOIN {} ON ({}. = {}.)'.format(join_to, sources[0],
+                                                           join_to))
+    return '\n'.join(from_clause)
+
+
+def cast(column_str, new_type, db_url):
+    'Produces a version of `column_str` cast to data type `new_type` in `db_url` dialtect'
+    engine_name = db_engine_name(db_url)
+    if engine_name == 'postgresql':
+        return '{}::{}'.format(column_str, new_type)
+    elif engine_name == 'mysql':
+        return 'CAST({} AS {})'.format(column_str,
+                                       new_type)  # TODO: include data length
```

Comment:
How urgent is fixing this TODO? Should it be filed as an issue?

Reply:
1. Not, but 2. Yes, I will.